services/gitlab/gitlab-contributors.service.js

```diff
@@ -0,0 +1,86 @@
+import Joi from 'joi'
+import parseLinkHeader from 'parse-link-header'
+import { optionalUrl } from '../validators.js'
+import { renderContributorBadge } from '../contributor-count.js'
+import GitLabBase from './gitlab-base.js'
+
+// check response type
+const schema = Joi.array().items(Joi.object())
+
+const queryParamSchema = Joi.object({
+  gitlab_url: optionalUrl,
+}).required()
+
+const documentation = `
+<p>
+  You may use your GitLab Project Id (e.g. 278964) or your Project Path (e.g. gitlab-org/gitlab )
+</p>
+`
+
+const customDocumentation = `
+<p>
+  Note that only network-accessible jihulab.com and other self-hosted GitLab instances are supported.
+  You may use your GitLab Project Id (e.g. 13953) or your Project Path (e.g. gitlab-cn/gitlab ) in <a href="https://jihulab.com">https://jihulab.com</a>
+</p>
+`
+
+export default class GitlabContributors extends GitLabBase {
+  static category = 'activity'
+  static route = {
+    base: 'gitlab/v/contributor',
+    pattern: ':project+',
+    queryParamSchema,
+  }
+
+  static examples = [
+    {
+      title: 'GitLab contributors',
+      namedParams: {
+        project: 'gitlab-org/gitlab',
+      },
+      staticPreview: this.render({ contributorCount: 418 }),
+      documentation,
+    },
+    {
+      title: 'GitLab (custom server) contributors',
+      queryParams: { gitlab_url: 'https://jihulab.com' },
+      namedParams: {
+        project: 'gitlab-cn/gitlab',
+      },
+      staticPreview: this.render({ contributorCount: 415 }),
+      documentation: customDocumentation,
+    },
+  ]
+
+  static defaultBadgeData = { label: 'contributors' }
+
+  static render({ contributorCount }) {
+    return renderContributorBadge({ contributorCount })
+  }
+
+  async handle({ project }, { gitlab_url: baseUrl = 'https://gitlab.com' }) {
+    // https://docs.gitlab.com/ee/api/repositories.html#contributors
+    const { res, buffer } = await this._request({
+      url: `${baseUrl}/api/v4/projects/${encodeURIComponent(
+        project
+      )}/repository/contributors`,
+      options: { searchParams: { page: '1', per_page: '1' } },
+      errorMessages: {
+        404: 'project not found',
+      },
+    })
+    const parsed = parseLinkHeader(res.headers.link)
+    let contributorCount
+    if (parsed === null) {
+      const json = this._parseJson(buffer)
+      this.constructor._validate(json, schema)
```

Comment:
I think you've got the wrong idea here.

If we're not using any values from the body, there is no reason to validate the contents of the body.

The purpose of validating is to make sure the value we need exists and makes sense so if we get a response that doesn't have an `x-total` header we aren't trying to call `metric(undefined)` or something like that.

So we could write something like

```js
import { nonNegativeInteger } from '../validators.js'

const schema = Joi.object({ 'x-total': nonNegativeInteger }).required()

async handle(...
  // ...

  const data = this.constructor._validate(res.headers, schema)
  const contributorCount = data['x-total']
  return this.constructor.render({ contributorCount })
}
```

here to

- make sure the headers contain the value we want
- it is in the expected format

There are some docs on how and why we validate at https://github.com/badges/shields/blob/master/doc/input-validation.md hopefully that helps.

Reply:
Oh, I did misunderstand, I thought I needed to validate the response content. Thanks for your guidance, I will fix it.